File: module-git/src/main/java/com/timecat/module/git/sgit/activities/delegate/actions/FetchAction.java
package com.timecat.module.git.sgit.activities.delegate.actions;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;

import com.timecat.module.git.R;
import com.timecat.module.git.sgit.activities.RepoDetailActivity;
import com.timecat.module.git.sgit.database.models.Repo;
import com.timecat.module.git.sgit.dialogs.DummyDialogListener;
import com.timecat.module.git.sgit.repo.tasks.repo.FetchTask;

import java.util.ArrayList;

public class FetchAction extends RepoAction {
    public FetchAction(Repo repo, RepoDetailActivity activity) {
        super(repo, activity);
    }

    @Override
    public void execute() {
        fetchDialog().show();
        mActivity.closeOperationDrawer();
    }

    private void fetch(String[] remotes) {
        final FetchTask fetchTask = new FetchTask(remotes, mRepo, mActivity.new ProgressCallback(R.string.git_fetch_msg_init));
        fetchTask.executeTask();
    }

    private Dialog fetchDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
        final String[] originRemotes = mRepo.getRemotes().toArray(new String[0]);
        final ArrayList<String> remotes = new ArrayList<>();
        return builder.setTitle(R.string.git_dialog_fetch_title)
                .setMultiChoiceItems(originRemotes, null, new DialogInterface.OnMultiChoiceClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int index, boolean isChecked) {
                        if (isChecked) {
                            remotes.add(originRemotes[index]);
                        } else {
                            for (int i = 0; i < remotes.size(); ++i) {
                                if (remotes.get(i) == originRemotes[index]) {
                                    remotes.remove(i);
                                }
                            }
                        }
                    }
                })
                .setPositiveButton(R.string.git_dialog_fetch_positive_button, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        fetch(remotes.toArray(new String[0]));
                    }
                })
                .setNeutralButton(R.string.git_dialog_fetch_all_button, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        fetch(originRemotes);
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DummyDialogListener())
                .create();
    }
}

File: module-git/src/main/java/com/timecat/module/git/sgit/activities/delegate/actions/FetchAction.kt
package com.timecat.module.git.sgit.activities.delegate.actions

import android.app.AlertDialog
import android.app.Dialog
import com.timecat.module.git.R
import com.timecat.module.git.sgit.activities.RepoDetailActivity
import com.timecat.module.git.sgit.database.models.Repo
import com.timecat.module.git.sgit.dialogs.DummyDialogListener
import com.timecat.module.git.tasks.FetchTask

class FetchAction(repo: Repo, activity: RepoDetailActivity) : RepoAction(repo, activity) {
    override fun execute() {
        fetchDialog().show()
    }

    private fun fetch(remotes: Array<String>) {
        val fetchTask = FetchTask(remotes, mRepo, mActivity.ProgressCallback(R.string.git_fetch_msg_init))
        fetchTask.executeTask()
    }

    private fun fetchDialog(): Dialog {
        val builder = AlertDialog.Builder(mActivity)
        val originRemotes = mRepo.getRemotes().toTypedArray()
        val remotes = ArrayList<String>()
        return builder.setTitle(R.string.git_dialog_fetch_title)
            .setMultiChoiceItems(originRemotes, null) { dialogInterface, index, isChecked ->
                if (isChecked) {
                    remotes.add(originRemotes[index])
                } else {
                    for (i in remotes.indices) {
                        if (remotes[i] === originRemotes[index]) {
                            remotes.removeAt(i)
                        }
                    }
                }
            }
            .setPositiveButton(R.string.git_dialog_fetch_positive_button) { dialogInterface, i -> fetch(remotes.toTypedArray()) }
            .setNeutralButton(R.string.git_dialog_fetch_all_button) { dialogInterface, i -> fetch(originRemotes) }
            .setNegativeButton(android.R.string.cancel, DummyDialogListener())
            .create()
    }
}